test/UnitTest/Microsoft.AspNet.OData.Test.Shared/Builder/StructuralTypeConfigurationTest.cs

```diff
@@ -54,5 +56,73 @@ public void AddDynamicPropertyDictionary_ThrowsIfTypeIsNotDictionary()
                 string.Format("The argument must be of type '{0}'.", "IDictionary<string, object>"));
         }
+
+        /// <summary>
+        /// Tests the namespace assignment logic to ensure that user assigned namespaces are honored during registration.
+        /// </summary>
+        [Fact]
+        public void NamespaceAssignment_AutoAssignsNamespaceToStructuralType_AssignedNamespace()
+        {
+            // Act.
+            string expectedNamespace = "TestingNamespace";
+            ODataConventionModelBuilder modelBuilder = new ODataConventionModelBuilder()
+            {
+                Namespace = expectedNamespace
+            };
+
+            modelBuilder.EntitySet<Client>("clients");
+            modelBuilder.ComplexType<ZipCode>();
+
+            // Assert
+            Assert.Equal(expectedNamespace, modelBuilder.EntityType<Client>().Namespace);
+            Assert.Equal(expectedNamespace, modelBuilder.EntityType<MyOrder>().Namespace);
+            Assert.Equal(expectedNamespace, modelBuilder.EntityType<OrderLine>().Namespace);
+            Assert.Equal(expectedNamespace, modelBuilder.EntityType<OrderHeader>().Namespace);
+            Assert.Equal(expectedNamespace, modelBuilder.ComplexType<ZipCode>().Namespace);
+        }
+
+        /// <summary>
+        /// Tests the namespace assignment logic to ensure that default CLR namespace is used if user does not assign one.
+        /// </summary>
+        [Fact]
+        public void NamespaceAssignment_AutoAssignsNamespaceToStructuralType_DefaultNamespace()
+        {
+            // Act.
+            ODataConventionModelBuilder modelBuilder = new ODataConventionModelBuilder();
+
+            modelBuilder.EntitySet<Client>("clients");
+            modelBuilder.ComplexType<ZipCode>();
+
+            // Assert
+            Assert.Equal(typeof(Client).Namespace, modelBuilder.EntityType<Client>().Namespace);
+            Assert.Equal(typeof(MyOrder).Namespace, modelBuilder.EntityType<MyOrder>().Namespace);
+            Assert.Equal(typeof(OrderLine).Namespace, modelBuilder.EntityType<OrderLine>().Namespace);
+            Assert.Equal(typeof(OrderHeader).Namespace, modelBuilder.EntityType<OrderHeader>().Namespace);
+            Assert.Equal(typeof(ZipCode).Namespace, modelBuilder.ComplexType<ZipCode>().Namespace);
+        }
+
+        /// <summary>
+        /// Tests the namespace assignment logic to check the order in which namespace is assigned matters.
+        /// </summary>
+        [Fact]
+        public void NamespaceAssignment_AutoAssignsNamespaceToStructuralType_NamespaceAssignedAfterAddingEntities()
+        {
+            // Act.
+            string expectedNamespace = "TestingNamespace";
+            ODataConventionModelBuilder modelBuilder = new ODataConventionModelBuilder();
+
+            modelBuilder.EntitySet<Client>("clients");
+
+            modelBuilder.Namespace = expectedNamespace;
+            modelBuilder.ComplexType<ZipCode>();
```

Comment:
>modelBuilder.ComplexType<ZipCode>(); [](start = 12, length = 36)

Can we add another test case, for example:

1. Set the namespace at builder,
2. but set another namespace for certain type?

Reply:
Added a new test NamespaceAssignment_AutoAssignsNamespaceToStructuralType_AssignedNamespaceAndClassNamespace